### PR TITLE
daml ledger: implementation of ledger commands via json api

### PR DESCRIPTION
**Pull request by robin-da**
Implementation of the remaining daml ledger commands via HTTP JSON service
endpoints instead of gRPC. All commands are tested in Test/Ledger.hs.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs [hunk 1]
```diff
@@ -177,17 +177,28 @@ runLedgerUploadDar flags darPathM = do
       getDarPath
   putStrLn $ "Uploading " <> darPath <> " to " <> showHostAndPort args
   bytes <- BS.readFile darPath
-  uploadDarFile args bytes `catch` \(e :: SomeException) -> do
```
Comment (hurryabit) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs:180]: Is the only way `uploadDarFile` could fail in the past through the `fail` in `either fail return` below or were the other ways? If there were other ways, we wouldn't catch these exceptions anymore.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs [hunk 2]
```diff
@@ -177,17 +177,28 @@ runLedgerUploadDar flags darPathM = do
...
+uploadDarFile :: LedgerArgs -> BS.ByteString -> IO (Either String ())
+uploadDarFile args bytes =
+  case api args of
+    Grpc -> runWithLedgerArgs args $ do L.uploadDarFile bytes
+    HttpJson -> do
+      (i :: Int) <- httpJsonRequest args "POST" "/v1/packages" $ setRequestBodyLBS (BSL.fromStrict bytes)
+      return $
+        if i == 1
+          then Right ()
+          else Left "An error occured. Please check the returned status."
```
Comment (hurryabit) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs:201]: What is the "returned status"?

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs [hunk 3]
```diff
@@ -274,16 +285,25 @@ downloadAllReachablePackages args pid = loop Map.empty [pid]
...
+          mbPkg <- L.getPackage lid $ convPid pid
+          case mbPkg of
+            Nothing -> fail $ "Unable to download package with identity: " <> show pid
+            Just (L.Package bs) -> pure bs
+      HttpJson ->
+        httpBsRequest
+          args
+          "GET"
+          (Path $ unPath "/v1/packages/" <> (T.encodeUtf8 $ LF.unPackageId pid))
+          id
```
Comment (hurryabit) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs:302]: How are the exceptions that could happen here being reported?

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs [hunk 4]
```diff
@@ -274,16 +285,25 @@ downloadAllReachablePackages args pid = loop Map.empty [pid]
...
-        Left err -> fail $ show err
-        Right pkg -> return pkg
+  bs <-
+    case api args of
+      Grpc -> do
+        runWithLedgerArgs args $ do
+          lid <- L.getLedgerIdentity
+          mbPkg <- L.getPackage lid $ convPid pid
+          case mbPkg of
+            Nothing -> fail $ "Unable to download package with identity: " <> show pid
```
Comment (hurryabit) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs:295]: Calling `fail` is not the most user-friendly way to report errors. I'm absolutely aware that it was like this before but I think we can do better. Let's do that in a separate PR, if you don't mind.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs [hunk 5]
```diff
@@ -385,25 +409,45 @@ data HttpJsonResponseBody a = HttpJsonResponseBody
...
+  req <- makeRequest args method path modify
+  resp <- runHttp httpJSON req
+  pure $ result resp
+
+httpBsRequest :: LedgerArgs -> Method -> Path -> (Request -> Request) -> IO BS.ByteString
+httpBsRequest args method path modify = do
+  req <- makeRequest args method path modify
+  runHttp httpBS req
+
+makeRequest :: LedgerArgs -> Method -> Path -> (Request -> Request) -> IO Request
```
Comment (hurryabit) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs:430]: I must admit that passing around functions to modify requests feels a bit strange. But I don't know anything less strange given the design of the HTTP library... 🤷

File: daml-assistant/daml-helper/test/DA/Daml/Helper/Test/Ledger.hs [hunk 6]
```diff
@@ -1,24 +1,41 @@
...
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.Text as T
 import System.Environment.Blank
 import System.FilePath
+import System.IO.Extra
 import System.Process
 import Test.Tasty
 import Test.Tasty.HUnit
 
+readDarMainPackageId :: FilePath -> IO String
```
Comment (hurryabit) [inline daml-assistant/daml-helper/test/DA/Daml/Helper/Test/Ledger.hs:26]: The `inspect-dar` command has to solve a similar problem. Maybe there's already a function for this purpose?

**Review by hurryabit — APPROVED** [review]
Great work. Thank you so much!

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs (hunk 5 above)
Comment (robin-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs:430]: :shrug: couldn't come up with something else either...

File: daml-assistant/daml-helper/test/DA/Daml/Helper/Test/Ledger.hs (hunk 6 above)
Comment (robin-da) [reply daml-assistant/daml-helper/test/DA/Daml/Helper/Test/Ledger.hs:26]: I exposed one in the inspect-dar code.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs (hunk 4 above)
Comment (robin-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs:295]: agreed.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs (hunk 2 above)
Comment (robin-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs:201]: The returned status will be reported by the http client itself for example: `daml-helper: user error (Request failed with error code Status {statusCode = 401, statusMessage = "Unauthorized"})`

Comment (hurryabit) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs:201]: I see. Thanks for the explanation.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs (hunk 3 above)
Comment (robin-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs:302]: The http client will return the returned status code and the status message by the server, which is often not really helpful sadly, but that's all information there is. That's still better than what the gRPC endpoint is doing, which doesn't return anything at all and just fails.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs (hunk 1 above)
Comment (robin-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs:180]: I changed this now to show the error caused by a `Left` return value which is caused by invalid arguments to the command, and catch exceptions and hint for the maximum upload size, since these are likely thrown by the network layer.
